Feature: arsn 76 support new bucket format

welansari opened the pull request:
- Add new `vFormat` attribut to new buckets
- Update object CRUD + List functions to support new bucket format
- Update tests to test both old and new bucket format

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -393,7 +419,8 @@ class MongoClientInterface {
         // eslint-disable-next-line
         objVal.versionId = versionId;
-        const vObjName = formatVersionKey(objName, versionId);
+        const vObjName = formatVersionKey(objName, versionId, params.vFormat);
+        const mObjName = formatMasterKey(objName, params.vFormat);
```
take the chance to use more descriptive name : the `m` or `v` prefix looks like a minor "type" indicator, when it actually is most important...

```suggestion
        const versionKey = formatVersionKey(objName, versionId, params.vFormat);
        const masterKey = formatMasterKey(objName, params.vFormat);
```

francoisferrand replied:
additionally, maybe it is be better to tweak the code a bit more, to compute both the versionKey and masterKey in one call?

since most parameters of both functions are actually the same, and both calls are made almost every time...

welansari replied:
> additionally, maybe it is be better to tweak the code a bit more, to compute both the versionKey and masterKey in one call?

I'm not really a fan of doing that, i think it'll just over complicate things. IMO we should keep them separate just to keep things intuitive

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -724,11 +755,14 @@ class MongoClientInterface {
-    getLatestVersion(c, objName, log, cb) {
+    getLatestVersion(c, objName, vFormat, log, cb) {
+        const mObjName = formatMasterKey(objName, vFormat);
+        let vObjName = formatVersionKey(objName, '', vFormat);
```
take the change to remove `VID_SEPPLUS` (not used anymore) ; and introduce an alternaive constant for this empty string (`VID_NONE` ?)

also would be nice to have a comment to explain the "no version than increase" : which breaks the abstraction, and works only because the version is added at the end of the key...

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -743,13 +743,25 @@ class MongoClientInterface {
+                        return next(errors.NoSuchKey);
+                    }
+                    if (doc.value.isPHD) {
+                        this.getLatestVersion(c, objName, vFormat, log, (err, value) => {
```
should this be `params.vFormat` ?

francoisferrand replied:
also, strange to do this call twice (and the first time without the `vFormat` parameter : which does not seem like a valid use of this function...)
is this maybe some issue with partial commit?

welansari replied:
> should this be params.vFormat ?

In the waterfall we first get the bucket vFormat then pass it to the next function, we don't have a vFormat in the params

welansari replied:
I'm not sure i see where we call it twice ?

francoisferrand replied:
twice at line 730-740 then line 749-750 :

```javascript
            if (doc.value.isPHD) {
                this.getLatestVersion(c, objName, log, (err, value) => {
                    [...]
                    if (doc.value.isPHD) {
                        this.getLatestVersion(c, objName, vFormat, log, (err, value) => {
```

welansari replied:
That's probably a bug in Github, if you go to this you'll see that there is only one
https://github.com/scality/Arsenal/blob/3194f6941acae5ef7a5ec310ec8adc87c778fc5f/lib/storage/metadata/mongoclient/MongoClientInterface.js#L730-L769

francoisferrand replied:
indeed, very weird...

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -1053,12 +1063,23 @@ class MongoClientInterface {
+                }
+            } else {
+                // eslint-disable-next-line no-param-reassign
+                newParams.gte = inc(range);
```
why change to `inc(range)` ?
if this was a bug, it should be fixed on other brahces as well...

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -1034,7 +1035,16 @@ class MongoClientInterface {
+        } else {
+            const masterStream = new MongoReadStream(c, _params[0], _params.mongifiedSearch);
+            const versionStream = new MongoReadStream(c, _params[1], _params.mongifiedSearch);
+            stream = new MergeStream(
```
how is this related?
this updates the implementation of an internal function (`internalListObject`), to support passing an array of parameters.... but not the place where it is called : so it actually has no effect...

* if not needed, don't add it ; and avoid splitting commits too much (e.g. code and test should ideally be in same commit ; updating the implementation of a function and making use of it, if not too big...)
* please write a descriptive commit message (https://cbea.ms/git-commit/)
* take the change to add documentation header to the function, even if missing, when you are changing its prototype

welansari replied:
In the `listObject` function, we generate the internal params of the listing algorithm here:
https://github.com/scality/Arsenal/blob/ad62d8b2e0fe9b4f1b4a5536fffa075026c83026/lib/storage/metadata/mongoclient/MongoClientInterface.js#L1131-L1132
Those params depend on the version of the algorithm (DelimiterMaster, DelimiterVersions) as well as the vFormat. In `v0` the params are in form of an object and in `v1` they are in form of an array of two objects where each object contain the filters to apply to the master and version stream.

This is the way it is implemented in the `metadata` backend, and to avoid changing everything i reused the code and classes already there (MergeStream)

francoisferrand replied:
ok, let's leave it like this then...

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -703,46 +732,53 @@ class MongoClientInterface {
-                params, log, cb);
-        } else if (params && params.versionId && params.repairMaster) {
-            return this.putObjectVerCase4(c, bucketName, objName, objVal,
+        return this.getBucketVFormat(bucketName, log, (err, vFormat) => {
```
this is much more significant than just "updating tests" !
please try to be more careful when splitting commits...

francoisferrand commented:
i think we are missing *unit tests* for the code which was added : e.g. each of the `update XXXX function` should typically include unit tests to validate all the new code path (e.g. each error case...). Note i am talking about _unit_ tests only here, e.g. validating the code is behaving as expected, returning the right errors in the right place, ... (and using mocks instead of relying on an actual mongodb deployment, contrary to the functional tests).

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -703,46 +732,53 @@ class MongoClientInterface {
-                params, log, cb);
-        } else if (params && params.versionId && params.repairMaster) {
-            return this.putObjectVerCase4(c, bucketName, objName, objVal,
+        return this.getBucketVFormat(bucketName, log, (err, vFormat) => {
```
Actually, I am wondering if this should be done in here.... It seems to me that reading the metastore to find the "type" of bucket is already done in Cloudserver : so IMHO it would be more consistant if it staid there only, and this function would just take the "vFormat" in argument...

Another argument for this approach is that cloudserver has likely alreday queried the metastore to retrieve information on the bucket, so this extra call to the backend here may actually be redundant in many/most case.

How is it implemented in other backeds? Do we have the same call to metastore from within the `putObject` method ?

welansari replied:
> How is it implemented in other backends?

Cloudserver doesn't have access to the vFormat attribut, the logic that deals with that is offloaded to the metadata backend, this is also the case for other backends. So Cloudserver does query the `metastore` to get bucket attributs but it never gets the vFormat.

> Do we have the same call to `metastore` from within the putObject method ?

In metadata it's a bit different, but as far as i understood yes we always query the db to get the vFormat at each request.

https://github.com/scality/MetaData/blob/613bc56a9f174dd4ba95f2e513dc648260bac3d9/lib/server/VersioningRequestProcessor.js#L407-L411
^ Here `replicateRequest` is used to replicate the putObject (and deleteObject) action among the servers. We get the vFormat each time we call it.

jonathan-gramain replied:
In metadata the bucket metadata is cached once retrieved, which means that we only get them from the db when they're not in the cache.

welansari replied:
Do you recommend adding a cache here? and if so how do we decide its sizing ?

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -703,46 +732,53 @@ class MongoClientInterface {
-            if (doc.value.isPHD) {
-                this.getLatestVersion(c, objName, log, (err, value) => {
+        async.waterfall([
+            next => this.getBucketVFormat(bucketName, log, next),
```
same here (and in the other methods)

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/arsenal/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-76 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.1.33`


Please check the `Fix Version/s` of ARSN-76, or the target
branch of this pull request.

francoisferrand commented on tests/functional/metadata/mongodb/getObject.spec.js:
```diff
@@ -106,6 +111,25 @@ describe('MongoClientInterface::metadata.getObjectMD', () => {
             });
     }
 
+    function setObjectVFormat(vFormat, cb) {
```
should not be needed: just create the bucket with the right format in its metadata?

welansari replied:
Bucket metadata is stored inside the `BucketInfo` class, and since it doesn't include vFormat we can't really do that

bert-e commented:
# Conflict




There is a conflict between your branch `feature/ARSN-76-support-new-bucket-format` and the
destination branch `development/8.1`.

Please resolve the conflict on **the feature branch** (`feature/ARSN-76-support-new-bucket-format`).

```sh
git fetch && \
git checkout origin/feature/ARSN-76-support-new-bucket-format && \
git merge origin/development/8.1
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:feature/ARSN-76-support-new-bucket-format
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -388,27 +421,38 @@ class MongoClientInterface {
+     * @param {Object} c bucket collection
+     * @param {String} bucketName bucket name
+     * @param {String} objName object name
+     * @param {Object} objVal object metadata
```
whenever possible, please specify the expected "type" of the Object, like:

```suggestion
     * @param {ObjectMD} objVal object metadata
```

welansari replied:
The `ObjectMD` example here might bring confusion cause what we get is an object that contains the ObjectMD data not the actual instance of the `ObjectMD` class

jonathan-gramain commented on lib/algos/list/skip.js:
```diff
@@ -59,10 +59,17 @@ class Skip {
+                }
                 /* Avoid to loop on the same range again and again. */
-                if (newRange === this.gteParams) {
+                if (JSON.stringify(newRange) === JSON.stringify(this.gteParams)) {
```
This is tricky, but you can actually leave this check alone (in v1, it would actually never be true in either case). For the context, the check was introduced in metadata to fix https://scality.atlassian.net/browse/MD-678.

I cannot tell with full confidence whether a similar issue would never happen with v1 actually, but clearly the fix here wouldn't work in v1 due to how the data is laid out internally, and here we compare the user param `gte` against the internal skip params which include the v1 special prefix. In metadata I did not change the original logic, presumably for that reason (I don't remember since it's been a while!)

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -189,21 +188,36 @@ class MongoClientInterface {
+        };
+        if (bucketName !== constants.usersBucket &&
+            bucketName !== PENSIEVE) {
+            payload.vFormat = BUCKET_VERSIONS.v1;
```
The vFormat for new buckets should be configurable at least (via an env var?), and even default to v0 maybe for the time being until we get confidence that it's reliable (it's still in beta phase for now in metadata, so likely the same should go for Mongo).

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -270,6 +284,25 @@ class MongoClientInterface {
+        }, {}, (err, doc) => {
+            if (err) {
+                log.error(
+                    'getBucketVFormat: error getting bucket vFormat',
```
We should log the bucket name here

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -668,68 +753,127 @@ class MongoClientInterface {
+            if (err) {
+                return cb(err);
+            }
+            if (!params) {
```
I'd suggest creating a new object `_params = Object.assign({}, params)` to avoid modifying the argument, since it's an API function we cannot assume that it's safe to modify `params`.

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -668,68 +753,127 @@ class MongoClientInterface {
+            next => this.getBucketVFormat(bucketName, log, next),
+            (vFormat, next) => {
+                if (params && params.versionId) {
+                    // eslint-disable-next-line
```
This eslint exception may not be needed anymore

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -668,68 +753,127 @@ class MongoClientInterface {
+                    // eslint-disable-next-line
+                    key = formatVersionKey(objName, params.versionId, vFormat);
+                } else {
+                    // eslint-disable-next-line
```
same

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -668,68 +753,127 @@ class MongoClientInterface {
     }
 
+    /**
+     * gets versionned and non versionned object metadata
```
```suggestion
     * gets versioned and non-versioned object metadata
```

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -668,68 +753,127 @@ class MongoClientInterface {
+                }, {}, (err, doc) => {
                     if (err) {
-                        log.error('getLatestVersion: getting latest version',
+                        log.error('findOne: error getting object',
```
The bucket and object key should be logged as well

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -668,68 +753,127 @@ class MongoClientInterface {
+                    if (doc.value.isPHD) {
+                        this.getLatestVersion(c, objName, vFormat, log, (err, value) => {
+                            if (err) {
+                                log.error('getLatestVersion: getting latest version',
```
Maybe time to add the bucket/object key here too

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -668,68 +753,127 @@ class MongoClientInterface {
+                    if (!doc) {
+                        return next(errors.NoSuchKey);
+                    }
+                    if (doc.value.isPHD) {
```
I think we could do `return next(null, doc)` and create a new step in the waterfall for the following logic

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -668,68 +753,127 @@ class MongoClientInterface {
+            filter.$gt = masterKey;
+            filter.$lt = lastVersionKey;
+        } else {
+            filter.$gt = versionKey;
```
It may not be enough to retrieve only the version key in v1, due to the fact that a bucket may be nonversioned for some time and only create master keys without associated version keys. I think we need to check both prefixes for that reason in v1.

welansari replied:
`getLatestVersion` only returns the versions and never the master, this is the case in previous implementation as well. You can see that it gets keys greater than the master key and not greater or equal.
If i changed the function to also check the master prefix it would mess up the other functions calling it. An example would be the repair function that gets the latest version to decide if it should repair or delete the master.
Having the function in it's current state doesn't create a problem when retrieving objects that don't have a version associated to them as we get them by their master key in `getObject`

jonathan-gramain replied:
OK I didn't notice that this function was for internal use, good point

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -751,20 +895,30 @@ class MongoClientInterface {
+     * @return {undefined}
      */
-    repair(c, bucketName, objName, objVal, mst, log, cb) {
+    repair(c, bucketName, objName, objVal, mst, vFormat, log, cb) {
```
In metadata implementation, v1 does not create any PHD key since the repair is done at time of deletion of a specific version, hence the repair code should not have to deal with v1. But if you decided to keep the PHD in v1 in the Mongo implementation, then you can ignore my comment.

vrancurel replied:
We could get rid of the PHDs in MongoClientInterface but I suggest we do that in a v2:
E.g.
- Use transactions to atomically update/delete MasterKey + LastVersionKey MD (the rationale is to have the MD up-to-date at all time in the MasterKey for O(1) access of the MD.

So let's make it simple and fix the DeleteMarkers first in v1 (and therefore keep PHD related concepts).

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -940,20 +1140,29 @@ class MongoClientInterface {
                             return cb(err);
                         }
-                        return cb();
+                        return cb(null);
```
Is this change needed

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -971,22 +1180,67 @@ class MongoClientInterface {
+            if (err) {
+                return cb(err);
+            }
+            if (!params) {
```
Same comment about creating a new `_params` object

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -1035,22 +1295,72 @@ class MongoClientInterface {
     }
 
+    /**
+     * lists versionned and non versionned objects in a bucket
```
```suggestion
     * lists versioned and non-versioned objects in a bucket
```

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -1035,22 +1295,72 @@ class MongoClientInterface {
+            // DelimiterVersions in v0 also returns the same thing
+            // in v1 however it returns two objects containing filters to use for two separate
+            // listing streams (master and version)
+            const extentionParams = extension.genMDParams();
```
```suggestion
            const extensionParams = extension.genMDParams();
```

jonathan-gramain commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -1035,22 +1295,72 @@ class MongoClientInterface {
-        internalParams.mongifiedSearch = params.mongifiedSearch;
-        return this.internalListObject(bucketName, internalParams, extension,
-            log, cb);
+        return this.getBucketVFormat(bucketName, log, (err, vFormat) => {
```
I'm not sure how this works with Mongo, but I'm surprised that we need to take into account the vFormat of the bucket to list its multipart-uploads. Normally (in metadata at least), listing multipart uploads goes through a special sidekick bucket that is just a flat list of open multipart uploads, hence always in v0. Can you check this to make sure it's correct?

welansari replied:
Well i didn't exclude the mpu buckets in the createBucket function so they too have a v0 and v1 key format, but you're right there's no point in having them support the v1 format, i'll fix that

jonathan-gramain commented on lib/storage/metadata/mongoclient/utils.js:
```diff
@@ -117,11 +120,79 @@ function translateConditions(depth, prefix, object, cond) {
+ * @return {String} formatted key
+ */
+function formatMasterKeyV0(key) {
+    return `${key}`;
```
```suggestion
    return key;
```

vrancurel commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -114,6 +114,11 @@ class MongoClientInterface {
             ? Number.parseInt(process.env.CONCURRENT_CURSORS, 10)
             : CONCURRENT_CURSORS;
+
+        this.bucketMetadataCache = new LRUCache(constants.maxCachedBuckets);
```
A value of 1000 shall be fine.
Can you monitor the cache/hit miss somewhere ?

Commit:
feature: ARSN-76 Added support for new bucket key format attribut

Bucket metadata now has a new attribut vFormat that stores
the version of key formatting used for the bucket's objects

Added utility helper functions that format the keys according to
the new attribut + unit tests

Added chache for the vFormat attribut, because of the additional
db calls that will be added

Updated putBucketAttributes so that it only updates the required
values otherwise it overwrites the vFormat

Added helper function that gets bucket vFormat

Commit:
feature: ARSN-76 Updated putObject functions

Updated putObject function to first get the bucket vFormat
attribut and pass it to internal functions

Updated putObjectVerCase functions to use formatting helper
functions to take into account the new vFormat bucket attribut

Updated getLatestVersion to take into account the new key formatting

Added unit tests & updated functional tests to use both key formats

Commit:
feature: ARSN-76 updated getObject function

Updated getObject function to get bucket vFormat
and use it to correctly format the object key

Added unit tests & updated functional tests to test
both formatting versions

Updated functional tests to use proper versions to avoid
confusion

Commit:
feature: ARSN-76 Updated deleteObject functions

Updated deleteObject function and it's internal
functions to support the new objet key format

Added unit tests + updated functional tests to test
both formatting versions

Commit:
feature: ARSN-76 Updated listObject function

Added support to listing algo returning parameters requesting
dual synchronized listings, as now for v1 buckets master and version
keys have different prefixes

Added unit tests and updated functional tests to list in both bucket formats

Updated listMultipartUploads to call the internal listing function with the
correct params

Commit:
feature: ARSN-76 Updated deleteObjectWithCond & putObjectWithCond functions

Updated functions to use the new object key format
Added unit test + updated functional tests to use vFormat

Commit:
feature: ARSN-76 Updated the MongoClientInterface unit tests

Updated tests to use the correct object keys when querying the db

vrancurel commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -148,6 +157,16 @@ class MongoClientInterface {
+            this.cacheHitMissLoggerInterval = setInterval(() => {
+                let hitRatio = (this.cacheHit / (this.cacheHit + this.cacheMiss)) || 0;
+                hitRatio = hitRatio.toFixed(3);
+                this.logger.info('MongoClientInterface: Bucket vFormat cache hit/miss (5min)',
```
`logger.debug` shall be enough I think

francoisferrand commented on lib/algos/list/skip.js:
```diff
@@ -59,8 +59,15 @@ class Skip {
             if (++this.streakLength >= MAX_STREAK_LENGTH) {
-                const newRange = this._inc(skippingRange);
-
+                let newRange;
```
can we add a (unit) test for this?

francoisferrand commented on lib/constants.js:
```diff
@@ -148,4 +150,10 @@ module.exports = {
+    maxCachedBuckets: process.env.METADATA_MAX_CACHED_BUCKETS ?
+        Number(process.env.METADATA_MAX_CACHED_BUCKETS) : 1000,
+    // default bucket key format version to use
+    defaultBucketVFormat: process.env.DEFAULT_BUCKET_KEY_FORMAT ?
```
not really a `constant` ; should this be added to the config instead?

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -114,6 +114,15 @@ class MongoClientInterface {
+        this.bucketVFormatCache = new LRUCache(constants.maxCachedBuckets);
+        this.defaultBucketKeyFormat = [BUCKET_VERSIONS.v0, BUCKET_VERSIONS.v1]
+            .includes(constants.defaultBucketVFormat) ? constants.defaultBucketVFormat
+            : BUCKET_VERSIONS.v0;
```
should this not default to v1 ?

btw, you got 2 places where the default value is defined : here and in the "constant" defintion. it would be better to have a single default value.

welansari replied:
it should yes, fixed that

jonathan-gramain replied:
For the context, in S3C we leave it to v0 for now as a default because we will introduce the feature as NFI, to make sure early adopters do not have issues before making it available to every customer.

For Artesca the context is different since there are far fewer customers and less stored data, and you want customers to benefit from v1 ASAP, so I think it is good to enable v1 by default, but just so you know.

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -148,6 +157,16 @@ class MongoClientInterface {
             });
             this.adminDb = client.db('admin');
+            // log cache hit/miss every 5min
+            this.cacheHitMissLoggerInterval = setInterval(() => {
```
should we make these (cachehHit/miss) available as prometheus metrics?
this would be easier to consume and monitor...
Should probably be done later, in a separate PR.

welansari replied:
it's out of the scope of this PR, but yes good idea

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -189,21 +209,37 @@ class MongoClientInterface {
+        if (bucketName !== constants.usersBucket &&
+            bucketName !== PENSIEVE &&
+            !bucketName.startsWith(constants.mpuBucketPrefix)) {
+            payload.vFormat = this.defaultBucketKeyFormat;
```
generally (and thinking about typing/upcoming typescript migration) : vFormat should be initialized anyway, e.g. to v0 format.
Or if we need an extra different behavior we can introduce yet another value (like "unversioned")

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -212,6 +248,8 @@ class MongoClientInterface {
                 return cb(errors.InternalError);
             }
+            // caching bucket vFormat
+            this.bucketVFormatCache.add(bucketName, payload.vFormat);
```
c.f. previous comment : `paylog.vFormat` may not be defined here.

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -268,6 +313,42 @@ class MongoClientInterface {
+            if (!doc) {
+                return cb(null, BUCKET_VERSIONS.v0);
+            }
+            const vFormat = doc.vFormat ? doc.vFormat : BUCKET_VERSIONS.v0;
```
```suggestion
            const vFormat = doc.vFormat || BUCKET_VERSIONS.v0;
```

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -309,8 +390,10 @@ class MongoClientInterface {
         }, {
-            _id: bucketName,
-            value: newBucketMD,
+            $set: {
```
what 's the difference (with vs w/o `$set`) ? was it a bug?

welansari replied:
without $set the vFormat gets removed

francoisferrand commented:
/wait

Some open question has been raised about "activation" of this feature: e.g. should we make the feature configurable (at Zenko level), and consequently what should be the default here (e.g. keep v0 for compatibility, or default to v1 as the new normal).

jbertran commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -753,24 +804,41 @@ class MongoClientInterface {
-        }
-        return this.putObjectNoVer(c, bucketName, objName, objVal,
-            params, log, cb);
+        const _params = Object.assign({}, params);
```
Consider `const _params = { ...params };` instead - no side-effect, no doc reading required, same cost in memory.

welansari replied:
Linter doesn't seem to like this one

jbertran commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -753,24 +804,41 @@ class MongoClientInterface {
+                return cb(err);
+            }
+            _params.vFormat = vFormat;
+            if (_params && _params.versioning && !_params.versionId) {
```
Interestingly, all the following function calls use the same arguments. You could split this into one function that returns the `putObjectStrategy`, in which you could early-return in a lot of cases and improve readability.

```js
function getPutObjectStrategy(params) {
    if (_params.versionId === '') {
        return this.putObjectVerCase2;
    }
    if (!_params.repairMaster)  {
        return this.putObjectVerCase3;
    }
    [...]
    return this.putObjectNoVer;
}
```

jbertran reviewed (commented):
Nice work! That's a lot of moving parts :clap:

welansari commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.1`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `stabilization/7.10.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.1`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `stabilization/7.10.3`



Please check the status of the associated issue ARSN-76.

Goodbye kerkesni.
